src/cast_core/src/actors/CrudActor.cpp

```diff
@@ -0,0 +1,615 @@
+#include <cast_core/actors/CrudActor.hpp>
+#include <chrono>
+#include <memory>
+
+#include <yaml-cpp/yaml.h>
+
+#include <mongocxx/client.hpp>
+#include <mongocxx/collection.hpp>
+
+#include <boost/log/trivial.hpp>
+
+#include <boost/log/trivial.hpp>
+#include <bsoncxx/json.hpp>
+#include <gennylib/Cast.hpp>
+#include <gennylib/context.hpp>
+
+#include <gennylib/ExecutionStrategy.hpp>
+#include <gennylib/MongoException.hpp>
+#include <gennylib/conventions.hpp>
+
+using BsonView = bsoncxx::document::view;
+using CrudActor = genny::actor::CrudActor;
+using DocGenerator = genny::value_generators::DocumentGenerator;
+
+namespace YAML {
+
+template <>
+struct convert<mongocxx::read_preference> {
+    using ReadPreference = mongocxx::read_preference;
+    using ReadMode = mongocxx::read_preference::read_mode;
+    static Node encode(const ReadPreference& rhs) {
+        Node node;
+        auto mode = rhs.mode();
+        if (mode == ReadMode::k_primary) {
+            node["ReadMode"] = "primary";
+        } else if (mode == ReadMode::k_primary_preferred) {
+            node["ReadMode"] = "primaryPreferred";
+        } else if (mode == ReadMode::k_secondary) {
+            node["ReadMode"] = "secondary";
+        } else if (mode == ReadMode::k_secondary_preferred) {
+            node["ReadMode"] = "secondaryPreferred";
+        } else if (mode == ReadMode::k_nearest) {
+            node["ReadMode"] = "nearest";
+        }
+        auto maxStaleness = rhs.max_staleness();
+        if (maxStaleness) {
+            node["MaxStalenessSeconds"] = genny::TimeSpec(*maxStaleness).count();
+        }
+        return node;
+    }
+
+    static bool decode(const Node& node, ReadPreference& rhs) {
+        if (!node.IsMap()) {
+            return false;
+        }
+        if (!node["ReadMode"]) {
+            // readPreference must have a read mode specified.
+            return false;
+        }
+        auto readMode = node["ReadMode"].as<std::string>();
+        rhs = mongocxx::read_preference{};
+        if (readMode == "primary") {
+            rhs.mode(ReadMode::k_primary);
+        } else if (readMode == "primaryPreferred") {
+            rhs.mode(ReadMode::k_primary_preferred);
+        } else if (readMode == "secondary") {
+            rhs.mode(ReadMode::k_secondary);
+        } else if (readMode == "secondaryPreferred") {
+            rhs.mode(ReadMode::k_secondary_preferred);
+        } else if (readMode == "nearest") {
+            rhs.mode(ReadMode::k_nearest);
+        } else {
+            return false;
+        }
+        if (node["MaxStalenessSeconds"]) {
+            auto maxStaleness = node["MaxStalenessSeconds"].as<int>();
```

Comment:
Per genny's conventions of having time-based units be a first-class concept, you'll want to have this be a `TimeSpec` rather than an `int`. See my later comment for handling `wtimeout`.